server/src/workspace/DocsDownloader.ts

```diff
@@ -0,0 +1,165 @@
+import { spawn } from 'child_process';
+import * as os from 'os';
+import * as readline from 'readline';
+import { join, extname, basename, dirname } from 'path';
+import { readdir, mkdir, access } from 'fs/promises';
+import { createWriteStream, unlink } from 'fs';
+import * as http from 'http';
+import * as https from 'https';
+import { eachLimit } from 'async';
+import { F_OK } from 'constants';
+
+export default class DocsDownloader {
+    private static readonly POL_REV_REGEX = /\(Rev. ([0-9a-fA-F]{9})\)/;
+    public commitId: string | null = null;
+
+    public constructor(private storageFsPath: string, private logging = true) { }
+
+    private log(...args: any[]) {
+        if (this.logging)
+        {console.log(...args);}
+    }
+
+    // TODO fix race conditions if multiple starts called (eg. from didChangeConfiguration)
+    public async start(workspaceRoot: string, moduleDirectory: string, commitId: string | null = null) {
+        if ((commitId === '' || commitId === null) && this.commitId) {
+            this.log('DocsDownloader', 'using existing commit', this.commitId);
+            return;
+        }
+
+        if (!commitId) {
+            this.log('DocsDownloader', 'finding commit from pol executable');
+            this.commitId = commitId = await this.getPolRevision(workspaceRoot);
+            this.log('DocsDownloader', 'found commit id', commitId);
+        } else {
+            this.log('DocsDownloader', 'using provided commitId', commitId);
+            this.commitId = commitId;
+        }
+
+        if (commitId) {
+            const _commitId = commitId;
+            const modules = await this.availableModules(moduleDirectory);
+            await eachLimit(modules, 5, (moduleName, cb) => {
+                this.downloadDoc(_commitId, moduleName)
+                    .then(_ => cb())
+                    .catch(e => cb(e));
+            });
+        }
+    }
+
+    public getXmlDocPath(moduleEmFile: string): string | null {
+        if (extname(moduleEmFile).toLowerCase() !== '.em') {
+            return null;
+        }
+
+        const { commitId: commitId } = this;
+
+        if (commitId) {
+            const moduleName = basename(moduleEmFile, extname(moduleEmFile));
+            const docFile = join(this.storageFsPath, commitId, `${moduleName}em.xml`);
+            this.log(`Returning docFile for ${moduleEmFile} => ${docFile}`);
+            return docFile;
+        }
+
+        return null;
+    }
+
+    private async downloadDoc(commitId: string, moduleName: string): Promise<void> {
+        const uri = `https://raw.githubusercontent.com/polserver/polserver/${commitId}/docs/docs.polserver.com/pol100/${moduleName}em.xml`;
```

Comment:
would it make sense to make the '/docs/docs.polserver.com/pol100/' configurable? Or at least the pol100 ;)

Reply:
I think it's best to keep as-is, and cross that bridge when we get there 😅